File: typescript/packages/bench/src/contracts/scheduleContract.ts
/** Recurrence. */

/** How often a periodic component runs. */
export enum Cadence {
    Hourly = "hourly",
    Daily = "daily",
    Weekly = "weekly",
    Monthly = "monthly",
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace Cadence {
    const SECONDS: Readonly<Record<Cadence, number>> = {
        [Cadence.Hourly]: 3600,
        [Cadence.Daily]: 86400,
        [Cadence.Weekly]: 604800,
        [Cadence.Monthly]: 2592000,
    };

    /** Nominal period length in seconds (a month is 30 days here). */
    export function seconds(cadence: Cadence): number {
        return SECONDS[cadence];
    }
}

/** Anything that runs on a cadence. */
export interface ScheduleContract {
    readonly cadence: Cadence;

    /** Next run instant, aligned to the cadence grid. */
    nextRunSeconds(now: number): number;
}

File: rust/crates/atelier-bench/src/contracts/schedule_contract.rs
//! Recurrence.

/// How often a periodic component runs.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Cadence {
    Hourly,
    Daily,
    Weekly,
    Monthly,
}

impl Cadence {
    /// Nominal period length in seconds (a month is 30 days here).
    pub fn seconds(&self) -> u64 {
        match self {
            Cadence::Hourly => 3_600,
            Cadence::Daily => 86_400,
            Cadence::Weekly => 604_800,
            Cadence::Monthly => 2_592_000,
        }
    }
}

/// Anything that runs on a cadence.
pub trait ScheduleContract {
    fn cadence(&self) -> Cadence;

    /// Next run instant, aligned to the cadence grid.
    fn next_run_seconds(&self, now: u64) -> u64 {
        let period = self.cadence().seconds();

        now - (now % period) + period
    }
}
